README.md
## About shFokus

shFokus is Bash script to make you focus and stay productive by blocking distracting websites such as Facebook, Youtube, Twitter etc. It does the job by modifying your HOSTS file. So when you try to open those distracting websites it will go nowhere.

## Requirements

- Unix or Unix like based OS such as MacOS or Linux. (Windows is not tested)
- Bash version >= 3.2.57

## Installation

Method 1 - Clone the project repository from GitHub.

```
$ git clone git@github.com:rioastamal/shfokus.git
```

Method 2 - Download the zipped version from GitHub.

```
$ curl -O -L 'https://github.com/rioastamal/shfokus/archive/master.zip'
```

And then extract to your desired directory.

## Usage and Examples

Running shFokus with `-h` flag will give you list of options and examples.

```
$ bash ./shfokus.sh -h
```

```
Usage: ./shfokus.sh [OPTIONS]

Where OPTIONS:
  -a ACTION     Specify action name using ACTION.
  -f FILE       List of blocked websites from FILE. Default is ~/.shfokus. Will
                be created if not exists.
  -h            Print this help and exit.
  -r            Dry run mode. Print the content that will be written to hosts
                file.
  -t MINUTES    Specify how long in MINUTES sites should be blocked/unblocked.
                Default is '0' which means blocked/unblocked forever. Make sure
                there is no other shfokus instance running.
  -v            Display shFokus version.

List of available ACTION:
  - block    -> block website
  - unblock  -> release block

Example of ~/.shfokus (one domain per line):
  youtube.com
  www.youtube.com
  facebook.com
  www.facebook.com

Example 1 - Block distracting websites:

  $ sudo bash ./shfokus.sh -a block

Example 2 - Block distracting websites in 30 minutes:

  $ sudo bash ./shfokus.sh -a block -t 30

Example 3 - Unblock distracting websites.

  $ sudo bash ./shfokus.sh -a unblock

Example 4 - Unblock distracting websites in 10 minutes.

  $ sudo bash ./shfokus.sh -a unblock -t 10

-------------------------------- About shFokus --------------------------------

shFokus is Bash script to make you focus and stay productive by blocking
distracting websites such as Facebook, Youtube etc.

shFokus is free software licensed under MIT. Visit the project homepage
at https://github.com/rioastamal/shfokus.
```

### Create List of Distracting Websites

Bear in mind that this step is optional, if you skip this step shFokus will create it automatically. By default it will block Facebook, Youtube, Instagram, Twitter, Twitch and Netflix.

Let's start by creating list of websites that should be blocked. By default shFokus will search for `~/.shfokus` file to determine list of websites. In this example we are going to block Facebook and Twitter.

```
$ vi ~/.shfokus
```

```
facebook.com
www.facebook.coom
twitter.com
www.twitter.com
```

As you can see above you need to specify one domain per line.

### Block Distracting Websites

Because modifying HOSTS file `/etc/hosts` require root privilege then you need to run the command as root or with sudo. To start blocking use option `-a block`.

```
$ sudo bash ./shfokus.sh -a block
```

```
Stay focus and productive -- Have a great day :)
```

If you're concern about what shFokus about to write to your HOSTS file then you can run in dry-run mode first. By running in dry-run mode it will not touch your HOSTS file instead it will print to STDOUT.

```
$ bash ./shfokus.sh -a block -r
```

```
...
# Added by Docker Desktop
# To allow the same kube context to work on the host and the container:
127.0.0.1 kubernetes.docker.internal
# End of section
# shfokus_begin - DO NOT TOUCH
0.0.0.0 facebook.com www.facebook.coom twitter.com www.twitter.com
# shfokus_end
Stay focus and productive -- Have a great day :)
```

Notice that shFokus add new entry between `# shfokus_begin` and `# shfokus_end` at the end of HOSTS file.

### Block for Specific Duration

If you want block distracting websites for specific duration then use `-t` option. It accept value in minutes.

```
$ sudo bash ./shfokus.sh -a block -t 60
```

```
Stay focus and productive -- Have a great day :)
Block will be released in 3587 sec... CTRL+C to unblock
```

Block will automatically released in 60 minutes. If you want to release earlier use CTRL+C.

```
----- Interrupted by CTRL+C -----
Distracting websites has been unblocked -- Enjoy your day :)
```

### Unblock Distracting Websites

To release block of distracting websites use option `-a unblock`. It will remove values from your HOSTS file.

```
$ sudo bash ./shfokus.sh -a unblock
```

```
Distracting websites has been unblocked -- Enjoy your day :)
```

### Unblock Distracting Websites for Specific Duration

If you want unblock websites for specific duration then use `-t` option. Same as block it accept value in minutes.

```
$ sudo bash ./shfokus.sh -a unblock -t 10
```

```
Websites will be blocked in ${SECONDS_LEFT} sec... CTRL+C to block
```

Websites will be unblocked temporarily for 10 minutes. If you want to block earlier use CTRL+C.

```
----- Interrupted by CTRL+C -----
Stay focus and productive -- Have a great day :)
```

## Author

shFokus is written by Rio Astamal &lt;rio@rioastamal.net&gt;

## License

shFokus is open source licensed under [MIT license](http://opensource.org/licenses/MIT).


CHANGELOG.md
## CHANGELOG

### v2020-08-29

* New feature: Unblock with duration

### v2020-08-29

* Initial Release
* Block websites
* Block websites with duration
* Unblock websites

shfokus.sh
#!/bin/bash
#
# @author Rio Astamal <rio@rioastamal.net>
# @desc shFokus is Bash script to make you focus and stay productive by
#       blocking distracting websites suchs Facebook, Youtube etc.
# @url https://github.com/rioastamal/shfokus

readonly FOKUS_SCRIPT_NAME=$(basename $0)

FOKUS_VERSION="2020-09-17"
FOKUS_DRY_RUN="no"
FOKUS_MINUTES="0"

[[ -z "$FOKUS_FILE" ]] && FOKUS_FILE="~/.shfokus"
[[ -z "$HOSTS_FILE" ]] && HOSTS_FILE="/etc/hosts"

shfokus_help()
{
    echo "\
Usage: $0 [OPTIONS]

Where OPTIONS:
  -a ACTION     Specify action name using ACTION.
  -f FILE       List of blocked websites from FILE. Default is ~/.shfokus. Will
                be created if not exists.
  -h            Print this help and exit.
  -r            Dry run mode. Print the content that will be written to hosts
                file.
  -t MINUTES    Specify how long in MINUTES sites should be blocked/unblocked.
                Default is '0' which means blocked/unblocked forever. Make sure
                there is no other shfokus instance running.
  -v            Display shFokus version.

List of available ACTION:
  - block    -> block website
  - unblock  -> release block

Example of ~/.shfokus (one domain per line):
  youtube.com
  www.youtube.com
  facebook.com
  www.facebook.com

Example 1 - Block distracting websites:

  $ sudo bash ./shfokus.sh -a block

Example 2 - Block distracting websites in 30 minutes:

  $ sudo bash ./shfokus.sh -a block -t 30

Example 3 - Unblock distracting websites.

  $ sudo bash ./shfokus.sh -a unblock

Example 4 - Unblock distracting websites in 10 minutes.

  $ sudo bash ./shfokus.sh -a unblock -t 10

-------------------------------- About shFokus --------------------------------

shFokus is Bash script to make you focus and stay productive by blocking
distracting websites such as Facebook, Youtube etc.

shFokus is free software licensed under MIT. Visit the project homepage
at https://github.com/rioastamal/shfokus."
}

shfokus_see_help()
{
    echo "Try '$FOKUS_SCRIPT_NAME -h' for more information."
}

shfokus_ctrl_c()
{
    echo -e "\n"
    echo "----- Interrupted by CTRL+C -----"

    # Reset FOKUS_MINUTES so it will not spawn another timer
    FOKUS_MINUTES=0

    # Execute the 1st arg as it act as a function name
    $1

    exit 0
}

shfokus_block_in_minutes()
{
    # Catch CTRL+C and call unblock function
    trap 'shfokus_ctrl_c shfokus_unblock' INT

    # convert minutes to seconds
    local SECONDS_LEFT=$(( $FOKUS_MINUTES * 60 ))

    while (( $SECONDS_LEFT >= 0 ))
    do
        # clear line
        echo -ne "\033[K"
        echo -ne "\rBlock will be released in ${SECONDS_LEFT} sec... CTRL+C to unblock"
        sleep 1
        SECONDS_LEFT=$(( $SECONDS_LEFT - 1 ))
    done

    echo -e "\n"

    # Reset FOKUS_MINUTES so it will not spawn another timer
    FOKUS_MINUTES=0

    shfokus_unblock
}

shfokus_unblock_in_minutes()
{
    # Catch CTRL+C and call block function
    trap 'shfokus_ctrl_c shfokus_block' INT

    # convert minutes to seconds
    local SECONDS_LEFT=$(( $FOKUS_MINUTES * 60 ))

    while (( $SECONDS_LEFT >= 0 ))
    do
        # clear line
        echo -ne "\033[K"
        echo -ne "\rWebsites will be blocked in ${SECONDS_LEFT} sec... CTRL+C to block"
        sleep 1
        SECONDS_LEFT=$(( $SECONDS_LEFT - 1 ))
    done

    echo -e "\n"

    # Reset FOKUS_MINUTES so it will not spawn another timer
    FOKUS_MINUTES=0

    shfokus_block
}

shfokus_write_hosts_file()
{
    local MODE="block"
    [[ ! -z "$1" ]] && MODE="$1"

    # Content to write to HOSTS file
    local FOKUS_CONTENT=
    # Used for dry-run mode
    local DRY_RUN_FILE=$HOSTS_FILE

    read -r -d '' FOKUS_CONTENT << EOF
# shfokus_begin - DO NOT TOUCH
$( echo 0.0.0.0 $( <$FOKUS_FILE ) )
# shfokus_end
EOF

    [[ "$FOKUS_DRY_RUN" == "yes" ]] && DRY_RUN_FILE="/dev/stdout"

    [[ ! -w "$HOSTS_FILE" && "$FOKUS_DRY_RUN" != "yes" ]] && {
        echo "Error: Can not write to HOSTS file $DRY_RUN_FILE." >&2
        echo "Make sure it is writeable by running as root." >&2
        exit 403
    }

    # Output the content inside sub-shell because we are
    # reading and writing the same file
    echo "$(
      echo "$( <$HOSTS_FILE )" | \
      sed "/^# shfokus_begin/,/^# shfokus_end/d"
    )" > $DRY_RUN_FILE

    [[ "$MODE" == "unblock" ]] && {
        echo "Distracting websites has been unblocked -- Enjoy your day :)"

        [[ "$FOKUS_MINUTES" != "0" ]] && shfokus_unblock_in_minutes
        return 0
    }

    # Append our blocked site config at the end of Hosts file
    echo "${FOKUS_CONTENT}" >> $DRY_RUN_FILE
    echo "Stay focus and productive -- Have a great day :)"

    [[ "$FOKUS_MINUTES" != "0" ]] && shfokus_block_in_minutes
}

shfokus_block()
{
    shfokus_write_hosts_file "block"
}

shfokus_unblock()
{
    shfokus_write_hosts_file "unblock"
}

shfokus_init()
{
    # Issue for expanding "~" because treated as string literal
    # Simple hack is replacing "~" with the value of $HOME
    FOKUS_FILE=$( echo "$FOKUS_FILE" | sed "s#~#$HOME#" )

    local FOKUS_DIR="$( dirname $FOKUS_FILE )"
    [[ ! -w "$FOKUS_DIR" ]] && {
        echo "Error: Directory $FOKUS_DIR is not writeable. Make sure it is writeable by running as root." >&2
        exit 403
    }

    # Write shfokus file if it not exists
    [[ ! -f "$FOKUS_FILE" ]] && {
        cat <<EOF > $FOKUS_FILE
www.youtube.com
youtube.com
www.facebook.com
facebook.com
www.instagram.com
instagram.com
twitter.com
www.twitter.com
twitch.tv
www.twitch.tv
netflix.com
www.netflix.com
EOF
    }

    [[ ! -w "$FOKUS_FILE" ]] && {
        echo "Error: File $FOKUS_FILE is not writeable. Make sure it is writeable by running as root." >&2
        exit 403
    }
}

# Parse the arguments
while getopts a:f:hrt:v FOKUS_OPT;
do
    case $FOKUS_OPT in
        a)
            FOKUS_ACTION="$OPTARG"
        ;;

        f)
            FOKUS_FILE="$OPTARG"
        ;;

        h)
            shfokus_help
            exit 0
        ;;

        r)
            FOKUS_DRY_RUN="yes"
        ;;

        t)
            FOKUS_MINUTES="$OPTARG"
        ;;

        v)
            echo "shFokus version $FOKUS_VERSION"
            exit 0
        ;;

        \?)
            shfokus_see_help
            exit 400
        ;;
    esac
done

shfokus_init

case $FOKUS_ACTION in
    block)
        shfokus_block
    ;;

    unblock)
        shfokus_unblock
    ;;

    *)
        echo -e "Unrecognized action.\n" >&2
        shfokus_see_help
        exit 400
    ;;
esac

exit 0